refactor_show_prototype.jl

```diff
@@ -0,0 +1,136 @@
+using Pkg
+Pkg.activate(@__DIR__)
+using Latexify
+import Base: show
+using SnoopCompile
+
+
+const Lmime = MIME"text/latexify"
+const lmime = MIME("text/latexify")
+
+const opdict = (;
+    :* => :Multiplication,
+    :/ => :Division,
+    :+ => :Addition,
+    :- => :Subtraction,
+    :^ => :Pow,
+)
+for T in values(opdict)
+    println(T)
+end
+abstract type AbstractLatexifyOperation end
+for T in values(opdict)
+    @eval(
+        struct $(T){S} <: AbstractLatexifyOperation
+            args::S
+        end
+    )
+end
+
+const ops = (; (keys(opdict) .=> eval.(values(opdict)))...)
+ops[:*]([:a, :b])
+
+for op in pairs(ops)
+    println(op[1], op[2])
+    @eval latexop(::Val{:call}, ::Val{$(QuoteNode(op[1]))}, args) = $(op[2])(args)
+end
+
+
+abstract type AbstractLatexifyEnvironment end
+for T in [:Align, :Equation, :Inline, :Bracket, :Raw, :Table]
+    @eval(
+        struct $(T){S} <: AbstractLatexifyEnvironment
+            args::S
+        end
+    )
+end
+Align(x, y) = Align(hcat(x, y))
+
+
+
+latexop(ex::Expr) = latexop(Val(ex.head), Val(ex.args[1]), ex.args[2:end])
+latexop(x) = x
+latexop(x...) = x
+latexop(x::Matrix) = Align(x)
+latexop(::Val{:align}, x) = Align(x)
+latexop(::Val{:equation}, x) = Equation(x)
+latexop(lhs::Vector, rhs::Vector) = Align(lhs, rhs)
+
+show(io::IO, ::MIME"text/latexify", x) = print(io, x)
+
+compare_precedence(a, b) = Base.operator_precedence(a) > Base.operator_precedence(b)
```

Comment:
I'm not entirely sure this will do the right thing for things like `1 - (2 + 3 + 4)` -- here the precedence is equal, but `compare_precedences` will return `false`. Perhaps that is solved by changing the `>` to `>=`, but will it then do the right thing for `1 - (2 - 3 - 4)`?

I guess the `show` method for `:-` could give `:+` as precedence for left and `:-` for right.

Reply:
I agree, this will not always do the right thing

Reply:
I'm not even convinced that `operator_precedence` is reliably good for latex purposes. I think one might find plenty of cases where code would need parentheses while latex would not.

Reply:
If there exists such an ordering, it's pretty simple to implement your own `operator_precedence`. Otherwise one might need individual `compare_precedence(::Val, ::Val)` functions.

Reply:
individual comparisons is pretty much what I have been doing before (not saying that this it the way to go). It turned out not to be needed too frequently though. `-` and `^` turned out to be the ones which needed the most care in figureing out where parentheses are needed.